### ISSUE TITLE
Support Sync with Kubernetes Secrets [optional secrets-store-csi-driver feature]

### ISSUE THREAD

**Issue by kmadel** [GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp#37]
### TL;DR
Add support for syncing to Kubernetes Secrets via `secretObjects`. See https://github.com/kubernetes-sigs/secrets-store-csi-driver#optional-sync-with-kubernetes-secrets

This would allow easier migration of existing pod manifests.

### Design

**Proposal**
Azure example: https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/test/bats/tests/azure/azure_v1alpha1_secretproviderclass_ns.yaml 
Vault example: https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/test/bats/tests/vault/vault_synck8s_v1alpha1_secretproviderclass.yaml 

**Alternatives considered**
No alternative.

**Resources**
Examples:
- Azure example: https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/test/bats/tests/azure/azure_v1alpha1_secretproviderclass_ns.yaml 
- Vault example: https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/test/bats/tests/vault/vault_synck8s_v1alpha1_secretproviderclass.yaml 

- [Link to documentation](TODO)

**Comment by tam7t** [comment]
This should be possible currently, although we don't have an example of it yet:

```yaml
apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
kind: SecretProviderClass
metadata:
  name: app-secrets
spec:
  provider: gcp
  secretObjects:
  - secretName: foosecret
    type: Opaque
    data: 
    - objectName: good1.txt
      key: pwd
  parameters:
    secrets: |
      - resourceName: "projects/$PROJECT_ID/secrets/testsecret/versions/latest"
        fileName: "good1.txt"
```

This will mount the `testsecret` to file `good1.txt` and sync it to a K8S secret called `foosecret`.

**Comment by kmadel** [comment]
The missing piece for me was "fileName" getting mapped to "objectName" but I guess that happens [here](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/26c0a5700abf38d9cf8674acc2da29811d250ca0/controllers/secretproviderclasspodstatus_controller.go#L154).
Tested with the following and it worked when a pod was created using the `secretProviderClass`:

```yaml
apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
kind: SecretProviderClass
metadata:
  name: cbci-mc-secret-provider
  namespace: cloudbees-core
spec:
  provider: gcp
  secretObjects:
  - secretName: cbci-mc-secret
    type: Opaque
    data: 
    - objectName: token
      key: token
  parameters:
    secrets: |
      - resourceName: "projects/core-workshop/secrets/cbci-workshop-token/versions/latest"
        fileName: "token"
```

**Comment by tam7t** [comment]
Thank you for verifying. At a minimum we need to add an integration test + documentation for this.

We could also rename `fileName` within the `parameters` to match `objectName`, but I do feel like `fileName is more descriptive.

**Comment by wilhelmi** [comment]
Do you have an example of your deployment where this worked? I can mount the secret in a volume, but I can't get it to sync to a secret.

**Comment by tam7t** [comment]
hey @wilhelmi the `SecretProviderClass` above should work.

If it does not then I suggest taking a look at the logs of the `csi-secrets-store` daemonset, If you get a message like:

```
"failed to create Kubernetes secret" err="secrets is forbidden: User \"system:serviceaccount:kube-system:secrets-store-csi-driver\" cannot create resource \"secrets\" in API group \"\" in the namespace \"default\"" spc="default/app-secrets" pod="default/mypod" secret="default/foosecret" spcps="default/mypod-default-app-secrets"
```

Then you may have missed the install step from https://secrets-store-csi-driver.sigs.k8s.io/getting-started/installation.html#alternatively-deployment-using-yamls that allows the csi driver to create secrets:

```
# If using the driver to sync secrets-store content as Kubernetes Secrets, deploy the additional RBAC permissions
# required to enable this feature
kubectl apply -f deploy/rbac-secretprovidersyncing.yaml
```

**Comment by 616b2f** [comment]
Does this mean when we change the password in e.g. azure keyvault the corresponding k8s `Secret` object get also updated or is only created once the pod is created and never updated?

**Comment by 616b2f** [comment]
Never mind I found the answer here: https://secrets-store-csi-driver.sigs.k8s.io/topics/secret-auto-rotation.html

TL;DR: you need to aktivate alpha feature via `--enable-secret-rotation=true` flag.

**Comment by joariasl** [comment]
If someone is here searching a method to load multiple environment variables from a single Secret in Google Cloud Secret Manager that create a single attribute in the resultant Secret.
I've resolved this creating an script to replace the Docker image Entrypoint to load the environments variables from an env file format to the container’s PID 1 scope.
https://gist.github.com/joariasl/ed88c2dc556695064dff7d6e89975415

**Comment by aaaaahaaaaa** [comment]
Personally I find very cumbersome (and counter-intuitive) the necessity to start a pod in order a the secret to be created. I was hoping for a solution that easily bridges GCP Secret Manager and Kubernetes secrets, without affecting my existing workloads definitions. But I understand that's a limitation coming from the driver itself, not this provider.

**Comment by dargudear-google** [comment]
Community is working on this solution in https://github.com/kubernetes-sigs/secrets-store-sync-controller . (Currently in alpha)

This will help us to tackle the above issue.
